src/client.rs

```diff
@@ -499,6 +502,100 @@ impl<T: Read + Write> Client<T> {
     }
 }
 
+#[cfg(feature = "rsasl")]
+impl<T: Read + Write> Client<T> {
+
+    /// Authenticate with the server using the given custom SASLConfig to handle the server's
+    /// challenge.
+    ///
+    pub fn sasl_auth(
+        mut self,
+        config: ::std::sync::Arc<SASLConfig>,
+        mechanism: &Mechname,
```

Comment:
It's a little weird to me to take `&Mechname` here, for a couple of reasons.

First, it seems like `rsals` actually wants `&[Mechname]`, so shouldn't we take that?
Second, if it's possible to _not_ provide an initial mechanism, should we make that possible?
And third, if `Mechname` is `Copy` (which its name makes it sound like it is), why the `&`?

Reply:
I think it's actually entirely reasonable for `Client` to start auto-issuing a call to gather the server's capabilities at the start of every session and storing the result. I'd be happy to land that, though I think we should land it in a separate PR maybe? And then you could make use of those stored capabilities here and get rid of the `&[Mechname]` argument entirely! Or, I suppose, we could make it an `Option<&[Mechname]>` so that users _can_ override if they want, but can also say "no, just choose based on server capabilities".